exe-common/src/sync.rs

```diff
@@ -659,3 +660,404 @@ pub fn get_chain_state_at_end_of(
+            delegate_pk: leader_xpub,
+            cert: leader_xprv.sign::<()>(&[0; 64]),
+        };
+        boot_stakeholders.insert(StakeholderId::new(&leader_xpub), leader_boot_stakeholder);
+        let genesis_data = config::GenesisData {
+            genesis_prev: cardano::block::HeaderHash::new(
+                &[0; cardano::hash::Blake2b256::HASH_SIZE],
+            ),
+            epoch_stability_depth: 2160,
+            start_time: SystemTime::UNIX_EPOCH + Duration::from_secs(1548089245),
+            slot_duration: Duration::from_millis(20000),
+            protocol_magic: ProtocolMagic::from(1097911063), /* testnet */
+            fee_policy: fee::LinearFee::new(fee::Milli::integral(155381), fee::Milli::new(43, 946)),
+            avvm_distr: BTreeMap::new(),
+            non_avvm_balances: BTreeMap::new(),
+            boot_stakeholders: boot_stakeholders,
+        };
+        let storage = Arc::new(RwLock::new(
+            Storage::init(&StorageConfig::new(test_dir.get_path()))
+                .expect("can't create local storage"),
+        ));
+        let mut net = LocalTestingPeer::new(&genesis_data, leader_xprv);
+        let net_config = net::Config {
+            genesis: net
+                .make_boundary_block(&BlockDate::from(0 as EpochId))
+                .header()
+                .compute_hash(),
+            genesis_prev: genesis_data.genesis_prev.clone(),
+            epoch_stability_depth: genesis_data.epoch_stability_depth,
+            protocol_magic: genesis_data.protocol_magic,
+            epoch_start: 0,
+            peers: net::Peers::new(),
+        };
+
+        // Sync once to get blocks into storage
+        for _ in 0..initial_blocks {
+            net.create_block();
+        }
+        let tip = net.get_tip().unwrap();
+        net_sync_to(&mut net, &net_config, &genesis_data, storage.clone(), &tip)
+            .expect("sync failed");
+
+        // Fork and generate new blocks in a longer chain and sync again to trigger the rollback
+        let removed = net.fork(fork_depth);
+        for _ in 0..new_blocks {
+            net.create_block();
+        }
+        let tip = net.get_tip().unwrap();
+        net_sync_to(&mut net, &net_config, &genesis_data, storage.clone(), &tip)
+            .expect("sync failed");
+        // We need to sync again as the previous sync rolls back but does not sync to the new chain.
+        net_sync_to(&mut net, &net_config, &genesis_data, storage.clone(), &tip)
+            .expect("sync failed");
+
+        // Check new chain is storage (by design old one should be there too)
+        let storage_read = storage.read().unwrap();
+        for removed_hash in removed.iter() {
+            assert!(storage_read
+                .block_exists(&cardano_storage::types::header_to_blockhash(&removed_hash))
+                .unwrap());
+        }
+        for (new_fork_hash, _new_fork_block) in net.blocks.iter() {
```

Comment:
This is not always the correct behavior for all cases, so I will update it tomorrow, in addition to possibly covering other flows if possible.